### PR TITLE
ENT-2726 Bypass selection page for enrolment url triggered login

### PR DESCRIPTION
**Pull request by irfanuddinahmad**
**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
- [ ] Version pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)

### CODE REVIEW

File: enterprise/tpa_pipeline.py [hunk 1]
```diff
@@ -142,7 +144,9 @@ def handle_redirect_after_social_auth_login(backend, user):
 
     """
     enterprise_customers_count = EnterpriseCustomerUser.objects.filter(user_id=user.id).count()
-    if enterprise_customers_count > 1:
+    next_url = backend.strategy.session_get('next')
+    using_enrolment_url = re.match(r'\/enterprise\/.*\/course\/.*\/enroll', next_url)
```
Comment (muhammad-ammar) [inline enterprise/tpa_pipeline.py:148]: I don't think we need to escape forward slashes `/`

Comment (irfanuddinahmad) [reply enterprise/tpa_pipeline.py]: Removed

File: enterprise/tpa_pipeline.py [hunk 2]
```diff
@@ -142,7 +144,9 @@ def handle_redirect_after_social_auth_login(backend, user):
 
     """
     enterprise_customers_count = EnterpriseCustomerUser.objects.filter(user_id=user.id).count()
-    if enterprise_customers_count > 1:
+    next_url = backend.strategy.session_get('next')
+    using_enrolment_url = re.match(r'/enterprise/.*/course/.*/enroll', next_url)
```
Comment (mattdrayer) [inline enterprise/tpa_pipeline.py:148]: `using_enrolment_url` -> `using_enrollment_url`

File: tests/test_tpa_pipeline.py [hunk 3]
```diff
@@ -190,6 +191,7 @@ def test_social_auth_user_login_associated_with_one_enterprise(self, new_associa
         kwargs = {'new_association': new_association}
         backend = self.get_mocked_sso_backend()
         backend.name = backend_name
+        backend.strategy.session_get.return_value = 'not-an-enrolment-url'
```
Comment (mattdrayer) [inline tests/test_tpa_pipeline.py:194]: `not-an-enrolment-url` -> `not-an-enrollment-url`

File: tests/test_tpa_pipeline.py [hunk 4]
```diff
@@ -207,6 +209,71 @@ def test_social_auth_user_login_associated_with_one_enterprise(self, new_associa
...
+        Test that enterprise selection page is bypassed if socialAuth user is part of multiple enterprises
+        and uses an enrolment url for login
+        """
+        kwargs = {'new_association': new_association}
+        backend = self.get_mocked_sso_backend()
+        backend.name = backend_name
+        if using_enrolment_url:
+            backend.strategy.session_get.return_value = '/enterprise/12e87171-fb0a/course/course-v1:Test/enroll'
+        else:
+            backend.strategy.session_get.return_value = 'not-an-enrolment-url'
```
Comment (mattdrayer) [inline tests/test_tpa_pipeline.py:248]: `not-an-enrolment-url` -> `not-an-enrollment-url`

File: tests/test_tpa_pipeline.py [hunk 5]
```diff
@@ -207,6 +209,71 @@ def test_social_auth_user_login_associated_with_one_enterprise(self, new_associa
...
+            enterprise_customer=enterprise_customer,
+            user_id=self.user.id,
+            active=True
+        )
+        with mock.patch('enterprise.tpa_pipeline.get_enterprise_customer_for_running_pipeline') as fake_get_ec:
+            with mock.patch(
+                    'enterprise.tpa_pipeline.select_enterprise_page_as_redirect_url') as ent_page_redirect:  # pylint: disable=invalid-name
+                fake_get_ec.return_value = None
+                handle_enterprise_logistration(backend, self.user, **kwargs)
+                if new_association or not multiple_enterprise_switch or using_enrolment_url:
```
Comment (mattdrayer) [inline tests/test_tpa_pipeline.py:272]: `using-enrolment-url` -> `using-enrollment-url`

File: enterprise/tpa_pipeline.py [hunk 6]
```diff
@@ -142,7 +144,9 @@ def handle_redirect_after_social_auth_login(backend, user):
...
+    if (enterprise_customers_count > 1) and not using_enrolment_url:
```
Comment (mattdrayer) [inline enterprise/tpa_pipeline.py:149]: `using_enrolment_url` -> `using_enrollment_url`

File: tests/test_tpa_pipeline.py [hunk 7]
```diff
@@ -148,6 +148,7 @@ def test_social_auth_user_login_associated_with_multiple_enterprise(self,
         kwargs = {'new_association': new_association}
         backend = self.get_mocked_sso_backend()
         backend.name = backend_name
+        backend.strategy.session_get.return_value = 'not-an-enrolment-url'
```
Comment (mattdrayer) [inline tests/test_tpa_pipeline.py:151]: `not-an-enrolment-url` -> `not-an-enrollment-url`

File: tests/test_tpa_pipeline.py [hunk 8]
```diff
@@ -207,6 +209,71 @@ def test_social_auth_user_login_associated_with_one_enterprise(self, new_associa
...
+        (True, False, False, 'google-oauth2'),
+        (False, False, False, 'google-oauth2'),
+        (True, False, True, 'google-oauth2'),
+        (False, False, True, 'google-oauth2'),
+        (True, True, False, 'google-oauth2'),
+        (False, True, False, 'google-oauth2'),
+    )
+    @ddt.unpack
+    @mock.patch('enterprise.tpa_pipeline.is_multiple_user_enterprises_feature_enabled')
+    def test_bypass_enterprise_selection_page_for_enrolment_url_login(self,
```
Comment (mattdrayer) [inline tests/test_tpa_pipeline.py:232]: `test_bypass_enterprise_selection_page_for_enrolment_url_login` -> `test_bypass_enterprise_selection_page_for_enrollment_url_login`

File: tests/test_tpa_pipeline.py [hunk 9]
```diff
@@ -207,6 +209,71 @@ def test_social_auth_user_login_associated_with_one_enterprise(self, new_associa
...
+                                                                      using_enrolment_url,
```
Comment (mattdrayer) [inline tests/test_tpa_pipeline.py:233]: `using_enrolment_url` -> `using_enrolment_url`

File: tests/test_tpa_pipeline.py [hunk 10]
```diff
@@ -207,6 +209,71 @@ def test_social_auth_user_login_associated_with_one_enterprise(self, new_associa
...
+                                                                      backend_name,
+                                                                      multiple_enterprises_feature):
+        """
+        Test that enterprise selection page is bypassed if socialAuth user is part of multiple enterprises
+        and uses an enrolment url for login
+        """
+        kwargs = {'new_association': new_association}
+        backend = self.get_mocked_sso_backend()
+        backend.name = backend_name
+        if using_enrolment_url:
```
Comment (mattdrayer) [inline tests/test_tpa_pipeline.py:245]: `using_enrolment_url` -> `using_enrolment_url`

**Pull request merged**